airflow/providers/odbc/hooks/odbc.py

```diff
@@ -0,0 +1,220 @@
+
+    DEFAULT_SQLALCHEMY_SCHEME = 'mssql+pyodbc'
+    conn_name_attr = 'odbc_conn_id'
+    default_conn_name = 'odbc_default'
+    supports_autocommit = True
+
+    def __init__(
+        self,
+        *args,
+        database: Optional[str] = None,
+        driver: Optional[str] = None,
+        dsn: Optional[str] = None,
+        connect_kwargs: Optional[dict] = None,
+        sqlalchemy_scheme: Optional[str] = None,
+        **kwargs,
+    ):
+        """
+        :param args: passed to DbApiHook
+        :param database: database to use -- overrides connection ``schema``
+        :param driver: name of driver or path to driver. overrides driver supplied in connection ``extra``
+        :param dsn: name of DSN to use.  overrides DSN supplied in connection ``extra``
+        :param connect_kwargs: keyword arguments passed to ``pyodbc.connect``
+        :param sqlalchemy_scheme: Scheme sqlalchemy connection.  Default is ``mssql+pyodbc`` Only used for
+          ``get_sqlalchemy_engine`` and ``get_sqlalchemy_connection`` methods.
+        :param kwargs: passed to DbApiHook
+        """
+        super().__init__(*args, **kwargs)
+        self._database = database
+        self._driver = driver
+        self._dsn = dsn
+        self._conn_str = None
+        self._sqlalchemy_scheme = sqlalchemy_scheme
+        self._connection = None
+        self._connect_kwargs = connect_kwargs
+
+    @property
+    def connection(self):
+        """
+        ``airflow.Connection`` object with connection id ``odbc_conn_id``
+        """
+        if not self._connection:
+            self._connection = self.get_connection(getattr(self, self.conn_name_attr))
+        return self._connection
+
+    @property
+    def database(self):
+        """
+        Database provided in init if exists; otherwise, ``schema`` from ``Connection`` object.
+        """
+        return self._database or self.connection.schema
+
+    @property
+    def sqlalchemy_scheme(self):
+        """
+        Database provided in init if exists; otherwise, ``schema`` from ``Connection`` object.
+        """
+        return (
+            self._sqlalchemy_scheme or
+            self.connection_extra_lower.get('sqlalchemy_scheme') or
+            self.DEFAULT_SQLALCHEMY_SCHEME
+        )
+
+    @property
+    def connection_extra_lower(self):
+        """
+        ``connection.extra_dejson`` but where keys are converted to lower case.
+
+        This is used internally for case-insensitive access of odbc params.
+        """
+        return {k.lower(): v for k, v in self.connection.extra_dejson.items()}
+
+    @property
+    def driver(self):
+        """
+        Driver from init param if given; else try to find one in connection extra.
+        """
+        if not self._driver:
+            driver = self.connection_extra_lower.get('driver')
+            if driver:
+                self._driver = driver
+        return self._driver and self._driver.strip().lstrip('{').rstrip('}').strip()
```

Comment:
i am stripping it so that user could do `{ODBC Driver 17 for SQL Server}` or `ODBC Driver 17 for SQL Server` and either way will work.

this way, whether user has provided them or not, when i build the connection string, i add safely add the curly braces.

Reply:
this way your conn extra can look like this:
```
{
      "Driver": "ODBC Driver 17 for SQL Server",
      "ApplicationIntent": "ReadOnly",
      "TrustedConnection": "Yes"
}
```
(i.e. no braces)

If you build your conn_uri from env vars, this is especially helpful because in airflow conn_uri format you have to urlencode curly braces so with braces it would look like this:
```
'none://?Driver=%7BODBC+Driver+17+for+SQL+Server%7D&ApplicationIntent=ReadOnly&TrustedConnection=Yes'
```

But without them it could be tad cleaner and more user friendly:
`'none://?Driver=ODBC+Driver+17+for+SQL+Server&ApplicationIntent=ReadOnly&TrustedConnection=Yes'`

it's just easier for users without really any risk or cost as i see it